### PR TITLE
Add Critical Network Chains to the CLI and HTML Report

### PR DESCRIPTION
**Pull request by paullewis**
PTAL @deepanjanroy @paulirish 

I think we established some quirks with the times reported, but as far as putting out the content goes, this works.

## CLI
screen shot 2016-05-12 at 8 18 08 pm

## HTML Report
screen shot 2016-05-12 at 8 17 07 pm

### CODE REVIEW

File: formatters/partials/critical-network-chains.html [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+.vert {
+  background: url('data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+Cjxzdmcgd2lkdGg9IjE2cHgiIGhlaWdodD0iMjZweCIgdmlld0JveD0iMCAwIDE2IDI2IiB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiPgogICAgPCEtLSBHZW5lcmF0b3I6IFNrZXRjaCAzLjcuMiAoMjgyNzYpIC0gaHR0cDovL3d3dy5ib2hlbWlhbmNvZGluZy5jb20vc2tldGNoIC0tPgogICAgPHRpdGxlPnZlcnQ8L3RpdGxlPgogICAgPGRlc2M+Q3JlYXRlZCB3aXRoIFNrZXRjaC48L2Rlc2M+CiAgICA8ZGVmcz48L2RlZnM+CiAgICA8ZyBpZD0iUGFnZS0xIiBzdHJva2U9Im5vbmUiIHN0cm9rZS13aWR0aD0iMSIgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj4KICAgICAgICA8ZyBpZD0idmVydCIgZmlsbD0iI0Q4RDhEOCI+CiAgICAgICAgICAgIDxyZWN0IGlkPSJSZWN0YW5nbGUtMTM4IiB4PSI3IiB5PSIwIiB3aWR0aD0iMiIgaGVpZ2h0PSIyNiI+PC9yZWN0PgogICAgICAgIDwvZz4KICAgIDwvZz4KPC9zdmc+');
```
Comment (samccone) [inline formatters/partials/critical-network-chains.html]: base 64 encoded images are cool, however it makes updating assets really hard

File: src/gatherers/critical-network-chains.js [hunk 2]
```diff
@@ -140,36 +139,29 @@ class CriticalNetworkChains extends Gather {
...
+
+    // Note: Approximately,
+    // startTime: time when request was dispatched
+    // responseReceivedTime: either time to first byte, or time of receiving
+    //  the end of response headers
+    // endTime: time when response loading finished
+    nonTrivialChains = nonTrivialChains.map(chain => ({
+      urls: chain.map(request => request._url),
+      totalRequests: chain.length,
+      times: chain.map(request => ({
```
Comment (samccone) [inline src/gatherers/critical-network-chains.js]: why do we need to extract these props into a new object with the same key names?

If we are attempting to prevent mutation of the original data perhaps we could use Array.from or Object.assign to create a clone

File: src/gatherers/critical-network-chains.js [hunk 3]
```diff
@@ -140,36 +139,29 @@ class CriticalNetworkChains extends Gather {
...
+        startTime: request.startTime,
+        endTime: request.endTime,
+        responseReceivedTime: request.responseReceivedTime
+      })),
+      totalTimeBetweenBeginAndEnd:
+        (chain[chain.length - 1].endTime - chain[0].startTime) * 1000,
```
Comment (samccone) [inline src/gatherers/critical-network-chains.js]: can we pull these into named reducers instead of defined inline?

File: formatters/formatter.js [hunk 4]
```diff
@@ -34,14 +34,16 @@ class Formatter {
...
+      accessibility: require('./accessibility'),
+      criticalNetworkChains: require('./critical-network-chains')
     };
   }
 
   static _generateSupportedFormats() {
     const formatNames = Object.keys(this._formatters);
     this._supportedFormatsNames = formatNames.reduce((prev, format) => {
-      prev[format.toUpperCase()] = format;
+      const formatName = format.replace(/([A-Z])/g, '_$1').toUpperCase();
```
Comment (samccone) [inline formatters/formatter.js]: pull this regex out of the function body to avoid recompile on each call.

File: formatters/critical-network-chains.js [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  /**
+   * Converts the tree into an ASCII tree.
+   */
+  static _createURLTreeOutput(info) {
+    const urlTree = CriticalNetworkChains._createURLTree(info);
+
+    function write(node, depth, treeMarkers) {
+      return Object.keys(node).reduce((output, itemURL, currentIndex, arr) => {
```
Comment (samccone) [inline formatters/critical-network-chains.js]: a+ use of reduce

File: formatters/critical-network-chains.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        // Create the appropriate tree marker based on the depth of this
+        // node as well as whether or not it has children and is itself the last child.
+        const treeMarker = ancestorTreeMarker +
+            (isLastChild ? '┗━' : '┣━') +
+            (hasChildren ? '┳' : '━');
+
+        const parsedURL = CriticalNetworkChains.parseURL(itemURL);
+
+        // Return the previous output plus this new node, and recursively write its children.
+        return output + `${treeMarker} ${parsedURL.file} (${parsedURL.hostname})` +
```
Comment (samccone) [inline formatters/critical-network-chains.js]: output can just be another interpolation here, actually pretty sure all of this can be :dancer:  no need for +

File: formatters/partials/critical-network-chains.html (hunk 1 above)
Comment (paullewis) [inline formatters/partials/critical-network-chains.html]: We need a single file for people to share the reports.

File: src/audits/performance/first-meaningful-paint.js [hunk 7]
```diff
@@ -88,6 +89,10 @@ class FirstMeaningfulPaint extends Audit {
           return FirstMeaningfulPaint.generateAuditResult({
             value: result.score,
             rawValue: result.duration,
+            extendedInfo: {
+              formatter: Formatter.SUPPORTED_FORMATS.CRITICAL_NETWORK_CHAINS,
+              value: artifacts.criticalNetworkChains
```
Comment (deepanjanroy) [inline src/audits/performance/first-meaningful-paint.js]: I'm guessing this is not needed anymore?

File: src/gatherers/critical-network-chains.js (hunk 2 above)
Comment (paullewis) [inline src/gatherers/critical-network-chains.js]: This is @deepanjanroy's code, I just re-jigged it. I think we should consider this area up for grabs as we figure out its accuracy and so on.

File: formatters/critical-network-chains.js (hunk 6 above)
Comment (paullewis) [inline formatters/critical-network-chains.js]: Indentation gets broken when you interpolate multiline. It preserves all that whitespace.

File: src/audits/performance/first-meaningful-paint.js (hunk 7 above)
Comment (paullewis) [inline src/audits/performance/first-meaningful-paint.js]: Yeah I removed it, just forgot to push that commit.

File: formatters/formatter.js (hunk 4 above)
Comment (paullewis) [inline formatters/formatter.js]: It's only run once, but I hear you.

**Comment by deepanjanroy** [comment]
screen shot 2016-05-12 at 4 46 12 pm
Can we set a max length on the url so the domain name is always visible?

File: formatters/critical-network-chains.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+  /**
+   * Refactors the gatherer's request information into a tree.
+   */
+  static _createURLTree(info) {
+    return info.reduce((tree, item) => {
+      let node = tree;
+      item.urls.forEach((itemURL, index, arr) => {
+        if (!node[itemURL]) {
+          const isLastChild = (index === arr.length - 1);
+          node[itemURL] = isLastChild ? item.totalTimeBetweenBeginAndEnd : {};
```
Comment (deepanjanroy) [inline formatters/critical-network-chains.js]: Using the url as key can potentially lead to trouble. There could be two different requests with the same url. This doesn't happen often, but it's safer to use requestId as the key.

Comment (deepanjanroy) [inline formatters/critical-network-chains.js]: It feels a little silly that we create the whole tree in the gatherer, then extract the chains out of it and throw the tree away, and then recreate the tree again here. We should be able to pass through the whole tree from the gatherer. We can do this refactor in a separate PR though.

File: formatters/critical-network-chains.js (hunk 6 above)
Comment (samccone) [inline formatters/critical-network-chains.js]: ah word

**Comment by paulirish** [comment]
Enjoying this direction. 

* "Critical request chains" (rather than "network").
* We can link to https://developers.google.com/web/fundamentals/performance/critical-rendering-path/analyzing-crp in the HTML description.

File: src/gatherers/critical-network-chains.js (hunk 3 above)
Comment (paulirish) [inline src/gatherers/critical-network-chains.js]: on the same front. I'd prefer to have a few separate statements in here. easier to debug and walk through..

----

20 minutes later.... here's a PR: https://github.com/paullewis/lighthouse/pull/4

File: extension/app/pages/scripts/lighthouse-report.js [hunk 9]
```diff
@@ -21,76 +21,37 @@
```
Comment (paulirish) [inline extension/app/pages/scripts/lighthouse-report.js]: this file is now in .gitignore so you should wipe it.

File: formatters/partials/critical-network-chains.html (hunk 1 above)
Comment (samccone) [inline formatters/partials/critical-network-chains.html]: yeah, :) just an aside no biggie yet

File: src/gatherers/critical-network-chains.js (hunk 2 above)
Comment (samccone) [inline src/gatherers/critical-network-chains.js]: kk

**Comment by paulirish** [comment]
Since we've got twitter's perf team interested in this... Shall we get it landed?

**Comment by boopathi** [comment]
Wouldn't it be useful to also mention whether the resource loaded `async`, `sync` or `defer` if js ? Or is it not to avoid cluttering?

**Comment by paullewis** [comment]
> Since we've got twitter's perf team interested in this... Shall we get it landed?

Yes.

> Wouldn't it be useful to also mention whether the resource loaded async, sync or defer if js ? Or is it not to avoid cluttering?

I don't know that we get that in the network records at this stage. If we do I could color code the JS requests accordingly. I'll take a look.

File: formatters/formatter.js (hunk 4 above)
Comment (paullewis) [inline formatters/formatter.js]: Done.

**Comment by paulirish** [comment]
> I don't know that we get that in the network records at this stage. If we do I could color code the JS requests accordingly. I'll take a look.

You could color based on [resource priority](https://docs.google.com/document/u/1/d/1bCDuq9H1ih9iNjgzyAL0gpwNFiEP4TZS-YLRp_RuMlc/edit), which is the high level signal that differentiates async/defer scripts from others.

image

**Comment by paullewis** [comment]
> You could color based on resource priority, which is the high level signal that differentiates async/defer scripts from others.

We could, but since we filter requests by priority, are we going to see any async / defer requests?

**Comment by paullewis** [comment]
Done a push so folks can take a look; tests currently fail.

**Comment by paulirish** [comment]
> We could, but since we filter requests by priority, are we going to see any async / defer requests?

Ah true. No we won't.  :)

**Comment by paullewis** [comment]
Okay PTAL.

**Comment by paullewis** [comment]
FYI next steps: updating the HTML report to be more "interactive" and "exciting" and "fun" and "scintillating" and "enthralling" and "awesome".

**Comment by deepanjanroy** [comment]
The redirect issue needs to be patched up (https://github.com/paullewis/lighthouse/pull/5), but apart from that everything looks great to me!

**Comment by paullewis** [comment]
@deepanjanroy Merged the PR. Can you +1 me?

File: formatters/critical-request-chains.js [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        return html;
+
+      default:
+        throw new Error('Unknown formatter type');
+    }
+  }
+
+  static _traverse(tree, cb) {
+    function walk(node, depth, startTime) {
+      const children = Object.keys(node);
```
Comment (samccone) [inline formatters/critical-request-chains.js]:
```
for (k in node) {
} 
```

?

File: formatters/critical-request-chains.js [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        const child = node[id];
+        if (!startTime) {
+          startTime = child.request.startTime;
+        }
+
+        // Call the callback with the info for this child.
+        cb({
+          depth,
+          id,
+          node: child,
```
Comment (samccone) [inline formatters/critical-request-chains.js]: if we just call it node here and rename the walk parent param to not be node this gets a lot nicer

File: formatters/critical-request-chains.js [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+      }, '');
+    }
+
+    walk(tree, 0);
+  }
+
+  static _getLongestChainLength(tree) {
+    let longestChain = 0;
+    this._traverse(tree, opts => {
+      const depth = opts.depth;
```
Comment (samccone) [inline formatters/critical-request-chains.js]: just a nit:

the value of a ivar here is debatable :)

File: formatters/critical-request-chains.js [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    });
+
+    // Always return the longest chain + 1 because the depth is zero indexed.
+    return (longestChain + 1);
+  }
+
+  static _getLongestChainDuration(tree) {
+    let longestChainDuration = 0;
+    this._traverse(tree, opts => {
+      const duration = opts.chainDuration;
```
Comment (samccone) [inline formatters/critical-request-chains.js]: ditto ivar comment

File: formatters/critical-request-chains.js [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+   * Converts the tree into an ASCII tree.
+   */
+  static _createURLTreeOutput(tree) {
+    function write(opts) {
+      const node = opts.node;
+      const depth = opts.depth;
+      const treeMarkers = opts.treeMarkers;
+      let startTime = opts.startTime;
+
+      return Object.keys(node).reduce((output, id, currentIndex, arr) => {
```
Comment (samccone) [inline formatters/critical-request-chains.js]: an annotation or comment describing the shape of node here would be useful :)

**Comment by paulirish** [comment]
lgtm here. 

I've put it up on #369 for the squash and land.

**Comment by paulirish** [comment]
image

👍
